Add API reference template and guide (#252)
thegooddocsproject/templates: api-reference/v2/api-reference-template.md

> # API reference overview

Reviewer: ```suggestion
# {Product} reference
```
* Suggeston: We have made use of multiple H1 headings. Some doc formats and editors only allow one H1 heading, and we probably should follow that convention too. (The first H1 heading typically becomes the title for the page, and is picked up by auto TOCs, browser tab titles, and bookmarks.)
* Suggestion: This would mean that the first title probably shouldn't include "overview"?
* Suggestion: I wonder whether we should include "API" in the title, or whether "{Product} reference" is sufficient. If a software product doesn't have another type of reference page, I'd be inclined to only show the "{Product} reference", and only include "API" if distinguishing from another type of reference. Interested to hear what conventions are out there for this.
More reasoning: "API" makes the title longer, and we want to keep titles short, especially TOC auto generated from titles.
* From a refererence page, I'm assuming we should mention the product in the title? I'm undecided on whether it should be 1. "{Product} API reference" or 2. "API reference for {Product}". (Note, in a browser tab, you will only see the first few chars, and if the user has tabs open for the tutorial, the reference, the howto, ..." then it is likely more important to start with the doctype before the product name. (This is descriptive text worth including in the deep dive or guide.)

Author: Altered the document title to `API Reference`, and increased the level of all previous headings to avoid confusion.

{Before using this template, read [the accompanying guide](api-reference-guide.md) to this template}.

## Introduction

> The {product name} APIs are REST APIs that provides {xx features}.

Reviewer: * Suggestion: (you can ignore): I have a mild preferene for {product} over {product name}

> The {product name} APIs are REST APIs that provides {xx features}.

Reviewer: Thought: 
* I see that {x}, {xx} and {xxx} is being used as a variable, sometimes inside a bigger {comment block}.
* If we continue using {x} as a variable, we should be consistant in the number of x's we use.
* In some situations, "XXX" is used as a categorization of pornographic material, which could be a reason to not use xxx. An alternative is {abc}, or {xyz}.
* From a style guide perspective, I'm thinking that we should use a {descriptive-variable-name} instead.
* If the variable is inside a larger {description block} then I suggest we use curly brackets inside curly brackets, such as: {This is an explanation with a {variable-name} inside it.}
* We should also set a convention for what case we use for variable names. I'd err toward {kabab-case}, but we could also use {normal text with spaces} or {lowerCamelCase} or {snake_case} or other conventions.
* We should take these questions to our newly formed style committee for a decision.

Author: Kept {snake_case} in this document for consistency with the endpoint description. I agree that it should be discussed as a style guideline

> The {product name} APIs are REST APIs that provides {xx features}.

Reviewer: ```suggestion
This reference describes the Application Program Interfaces (APIs) that can be used to {access | customize | program} {product}'s {features | functionality}.
```
* Suggestion: We probably shouldn't assume that all readers understand the meaning of "API". It is jargon understood by a group of developers, but some readers might not be developers. So we should spell out its meaning on first use.
* Suggestion: We probably shouldn't list product features here.  To stay current, this list would need to be updated as the product develops, introducing a maintence overhead, and it is likely to get out-of-date over time.

## Base URL

The APIs referenced in this documentation use the following base URL:

```
https://api.example.com
```

## Authorization

> All requests to the API must be authenticated and authorized before they can access or manage resources on the {environment or platform}.

Reviewer: Thought: A question to take to the style committee. How should present options in a variable? Sometimes we may want "or" as part of the option, such as {optionA or optionB | optionA and optionB}.
* I've previously used: / such as {environment / platform}.
* We could use upper case OR: such as {environment OR platform}.
* My mild preference would be to use the booleen logic char | such as {environment | platform}.

> The {product name} API uses the **{auth_type}** authorization type.

Reviewer: ```suggestion
The {product name} API uses the `{auth_type}` authorization type.
```
* Suggestion: Maybe using the code-block format for {auth_type}.

> You can authenticate your requests by {using authentication type and method}.

Reviewer: ```suggestion
You can authenticate your requests by using {authentication type and method}.
```
Suggestion: Moved the "using" out of the variable.

For example:

```

> <EXAMPLE>

Reviewer: ```suggestion
{EXAMPLE}
```
* Issue: For consistency, suggest we use curly brakets for variables. In the past, we've noticed that <angle-bracket> varables won't render in some cases, which is another reason not to use them.

```

> API requests without valid authentication will fail with a status of {error code}.

Reviewer: ```suggestion
API requests without valid authentication will fail with a status of ```{error code}```.
```
* Suggeston: I'm guessing most references would use code font for the error code?

## Versioning

{This section is optional.}

> The {product name} APIs provide {x} versions of API endpoints...

Reviewer: ```suggestion
The {product} APIs provide {number} versions of API endpoints.

{Commentary about the versions supported.}
```
* Suggest expanding "..." to explain what should be in "..."

## Pagination

{This section is optional.}

> You can use {xx} to paginate the results returned by the {product name} APIs. If not specified, the default values are ...

Reviewer: ```suggestion
Due to the potential for very large result sets from API calls, responses {are | can be} returned as shorter pages.

Pagination can be customized using {pagination settings}. If not specified, the default values are {values}.
```
Suggest structure the language to explain what pagination is.

## Rate limiting and throttling

{This section is optional.}

> The {product name} APIs use a {xx} rate limiting strategy. The maximum number of requests allowed to access a {resource/endpoint/..} is {xx} requests per {time period}.

Reviewer: Suggested tweaks.
```suggestion
The {product} APIs use a {name} rate limiting strategy. The maximum number of requests allowed to access a {resource/endpoint/..} is {number} requests per {time period}.
```

If the number of requests exceeds that limit, an HTTP `429 Too Many Request` error is returned.

> ## Status code

Reviewer: Suggestion: Good to list the standard status codes. What about product specific, non-standard codes?
I suspect we should describe how product specific codes are handled. Maybe say. There probably should be a table summary here, possibly with a description of codes within in API specific call description.

Author: add a new section to list the customized error types

The {product name} APIs use standard HTTP response codes.

| Status code | Message           | Description   |
|-------------|-------------------|---------------|

> | 200 `OK`    | Request succeeds. | {description} |

Reviewer: ```suggestion
| `200 OK`    | Request succeeds. | {description} |
```
Suggestion: We should be consistant in whether the error code and description are in the same code-block font. (It is different above.) I have a mild preference for them both being in the code font together.

|             |                   |               |
|             |                   |               |

# {Resource name}

The {resource name} is used to {xxx}.

## Data model

| Attribute | Type   | Required/Optional | Description                |
|-----------|--------|-------------------|----------------------------|
| {id}      | string | REQUIRED          | {Unique identifier of xx.} |
| {name}    | string | Optional          | {Name of xx}.              |
|           |        |                   |                            |

## Example

```
<EXAMPLE>
```

## Endpoints

Use the following endpoints to interact with the {resource name} entities.

| Method | Endpoint name                           | Description           |
|--------|-----------------------------------------|-----------------------|
| POST   | {[Endpoint name A](link_to_endpoint_a)} | Creates a {resource}. |
| GET    | {[Endpoint name B](link_to_endpoint_b)} | Returns a {resource}. |
|        |                                         |                       |

# {Endpoint name}

> {Provide a one-line description of what the API does. Starts with a verb.}

Reviewer: Suggestion: We probably shouldn't assume the reader (a software developer) has a good understanding of what a "verb" is, so this description should be followed by an example.

## Endpoint

```bash
{METHOD} /{request_url}/{{path_parameter}}
```

## Description

This endpoint ...

{This paragraph is optional.} This endpoint has been deprecated. Use {the recommended endpoint} instead. For more information about how to migrate to {the recommended endpoint}, see [{the migration guide}](link).

{This paragraph is optional.} The maximum number of calls to this API endpoint is {xx} per minute. For more information about API rate limiting/throttling, see [{the topic}](example).

## Authorization

The [{authorization method}](#authorization) is required for each API request.

> (Optional) Calling this endpoint also requires the {xx} permission.

Reviewer: ```suggestion
{This paragraph is optional.} Calling this endpoint also requires the {xx} permission.
```
* Suggestion: We should be consistent in how we describe optional status. In particular, we have been using {curly brackets} for writing tips.

## Request schema

### Path parameters

{This section is optional.}

> | Path parameter | Type   | Required/Optional | Description                |

Reviewer: ```suggestion
| Path parameter | Type   | Required?         | Description                |
```
* Mild suggestion: Suggest reducing words in the title.
* This comment applies to the other places this is referenced in the template and guide.

> | Path parameter | Type   | Required/Optional | Description                |

Reviewer 2: If a path parameter is part of the endpoint path, you must provide a value?

Author: good point. I added a note to the guide - "path parameters are always required"

|----------------|--------|-------------------|----------------------------|

> | {id}           | string | REQUIRED          | {Unique identifier of xx.} |

Reviewer: ```suggestion
| {id}           | string | Required          | {Unique identifier of xx.} |
```
* Suggest that we be consistent in our use of case, and I err on using lower case rather than upper case, as I don't think there is value gained by emphasizing with the capitals.
